aptos-move/block-executor/src/scheduler.rs

```diff
@@ -198,15 +200,14 @@ impl ValidationStatus {
     }
 }
 
-pub struct Scheduler {
-    /// Number of txns to execute, immutable.
-    num_txns: TxnIndex,
+pub struct Scheduler<P: ?Sized> {
+    txn_provider: Arc<P>,
 
     /// An index i maps to indices of other transactions that depend on transaction i, i.e. they
     /// should be re-executed once transaction i's next incarnation finishes.
-    txn_dependency: Vec<CachePadded<Mutex<Vec<TxnIndex>>>>,
+    txn_dependency: HashMap<TxnIndex, CachePadded<Mutex<Vec<TxnIndex>>>>,
```

Comment:
@gelash any concern with `Vec->HashMap`? TPS doesn't seem to be affected.

Reply:
Same change below, and in `LastInputOutput`.

Reply:
why not overallocate indices and use the needed ones? (like counting), but I'll defer to @ziaptos here. In general, we wouldn't see the performance impact here without contended load (otherwise no dependencies), and probably microbenchmark (else it won't show regression below some point). Although it's conceivable it won't regress too much because these accesses happen once per txn execution-ish

Reply:
I think I'd also prefer over-allocating. Though pre-filled hashmap might be ok as well - in any case I'd avoid anything that might allocate memory.